compiler/backend/ccgcalls.nim

```diff
@@ -11,11 +11,12 @@
 ## included from cgen.nim
 
 proc canRaiseDisp(p: BProc; n: PNode): bool =
-  # we assume things like sysFatal cannot raise themselves
+  ## Returns whether the callee expression `n` evaluates to a procedural values
+  ## for which a call can exit via exceptional control-flow. When panics are
+  ## not enabled, raising a defect also to exceptional control-flow.
   if n.kind == nkSym and {sfNeverRaises, sfImportc, sfCompilerProc} * n.sym.flags != {}:
```

Comment:
So this is the problematic line.

All ref-based ops are compiler procs, so error flag clearing for ref destructors will never be done as the system assumes no raises.

Reply:
Not quite. Auto-generated type-bound operations don't have the `sfCompilerProc` flag included. They do have the `sfNeverRaises` flag included, if the analysis in `liftdestructors` yields that they never call a user-provided operation.

In your test snippet from above, replacing `var o = O()` with `var o = (ref O)()` would make the code generator emit the error flag clear/restore properly, as the auto-generated `ref`-destructor doesn't have the `sfNeverRaises` flag set.

It's important to note that enabling `--panics:on` would break this again, as then the `canRaise` analysis is used, which treats all routines marked with `sfGeneratedOp` (which is set for all lifted type-bound operations) as never raising. Even if this wouldn't be the case, lifted operations don't inherit the effects from the operations they use, meaning that the effect specification for the lifted `=destroy` hook from the example is currently `.raises: []`.

Reply:
Yea I skipped some steps here, but what I meant is that the generated `=destroy` calls `nimDecRef*` which is `compilerproc` and is assumed to not raise and/or interact with error flags.

This is really a pain to test and I believe there will be more code that rely on these assumptions in the codebase.

Reply:
> the generated `=destroy` calls `nimDecRef*` which is `compilerproc` and is assumed to not raise and/or intera

Yep, but even them not being `.compilerproc`s wouldn't have helped here. Well, it would have prevented the dead-lock, but only because the `sharedDealloc` call would have never been reached then.